### PR TITLE
SGR-Pixels mouse encoding (1016)

### PR DESCRIPTION
**Pull request by klamonte**
This is for #1457 .

* DECRQM/DECRPM work for applications to detect support.
* Pixels are passed as offsets through the mouse event so that x/y retain their text coordinate meaning.  For SGR-Pixels encoding the x+offset / y+offset is converted back into screen pixel coordinates.
* I do not have test coverage, I think?  But all tests pass.

### CODE REVIEW

**Comment by klamonte** [comment]
@wez

I have been trying to figure out why I picked offsets within a cell rather than just pixels. It fit very naturally in jexer that way, but for other terminals I just did pixel_x/y as you initially suggested.

I finally understood why: in a multihead setup with different cell sizes, offset is better because moving the mouse across the entire screen on one head will move it across all heads. During multihead there might be some small distortion right on the right/bottom edges (jitter) as the mouse travels across the screen and exceeds or skips locations on the other heads, but it's not bad in practice.

I have seen requests for wezterm to support different font sizes in different windows, and thought cell+offset might make it easier if that scenario is supported later on.

But I can also switch really to pixel_x/y if you would prefer.

File: term/src/input.rs [hunk 1]
```diff
@@ -50,7 +52,7 @@ pub struct MouseEvent {
 #[derive(Debug, Clone)]
 pub struct LastMouseClick {
     pub button: MouseButton,
-    position: (usize, i64),
+    position: (usize, i64, usize, usize),
```
Comment (wez) [inline term/src/input.rs:55]: Stylistically speaking, big tuples are hard to grok.

Would you mind turning this tuple into a struct, and having `LastMouseClick::new` and `LastMouseClick::add` use that type for the position info?

Something like this should make it difficult to misinterpret the meaning of the fields in a future pass over the code:

```rust
struct ClickPosition {
   column: usize,
   row: i64,
   x_pixel_offset: usize,
   y_pixel_offset: usize,
}
```

File: termwiz/src/input.rs [hunk 2]
```diff
@@ -1267,6 +1267,21 @@ impl InputParser {
                                         modifiers,
                                     }));
                                 }
+                                // AZL - I don't quite understand this
```
Comment (wez) [inline termwiz/src/input.rs:1270]: Ah, this is for terminal applications who are listening for mouse reports from the terminal.
I think it's important for the application to know that the x/y are pixels for this case.
Perhaps the simplest way to encode that is to introduce a `InputEvent::PixelMouse(MouseEvent)` variant further up in this file:

https://github.com/wez/wezterm/blob/main/termwiz/src/input.rs#L63-L66

and then switch to `PixelMouse` in the callback we issue below?

**Review by wez — COMMENTED** [review]
Thanks for this!

I did wonder about the offsets in my first pass through, but honestly they feel fine to me!
Structurally this looks good! have a stylistic suggestion, and then we also need to reconcile the termwiz input mouse event -- comments inline!

File: termwiz/src/input.rs (hunk 2 above)
Comment (klamonte) [reply termwiz/src/input.rs]: How does this look now?

File: wezterm-gui/src/termwindow/mouseevent.rs [hunk 3]
```diff
@@ -105,8 +116,24 @@ impl super::TermWindow {
                 let button = mouse_press_to_tmb(press);
 
                 let click = match self.last_mouse_click.take() {
-                    None => LastMouseClick::new(button, (x, y)),
-                    Some(click) => click.add(button, (x, y)),
+                    None => LastMouseClick::new(
```
Comment (wez) [inline wezterm-gui/src/termwindow/mouseevent.rs:119]: nit: you could:

```
let click_position = ClickPosition {
   ...
};
```

ahead of this block, and then use pass that down to the `new` and `add` calls to save a little bit of repetition.

File: wezterm-gui/src/termwindow/mouseevent.rs [hunk 4]
```diff
@@ -165,7 +192,7 @@ impl super::TermWindow {
         if let Some(item) = ui_item {
             self.mouse_event_ui_item(item, pane, y, event, context);
         } else {
-            self.mouse_event_terminal(pane, x, y, event, context);
+            self.mouse_event_terminal(pane, x, y, x_pixel_offset, y_pixel_offset, event, context);
```
Comment (wez) [inline wezterm-gui/src/termwindow/mouseevent.rs:195]: how bad is the fan out from passing `ClickPosition` down here in place of the separate coordinates?  Does it add much value?

**Review by wez — COMMENTED** [review]
I think this is looking good!  I have a minor nit and a question about further code golf, but otherwise this looks fine!

File: wezterm-gui/src/termwindow/mouseevent.rs (hunk 4 above)
Comment (klamonte) [reply wezterm-gui/src/termwindow/mouseevent.rs:195]: No extra value that I can see.  I can switch these to ClickPosition also.

Comment (klamonte) [reply wezterm-gui/src/termwindow/mouseevent.rs:195]: mouse_event_terminal isn't referenced anywhere else.  I can also make it not pub. ?

**Comment by wez** [comment]
Thanks!

**Pull request merged**